src/theory/sets/theory_sets_rels.cpp

```diff
@@ -1058,49 +1056,72 @@ void TheorySetsRels::check(Theory::Effort level)
   void TheorySetsRels::doPendingInfers()
   {
     // process the inferences in d_pending
-    if (!d_sets_theory.isInConflict())
+    if (!d_state.isInConflict())
     {
-      std::vector<Node> lemmas;
       for (const Node& p : d_pending)
       {
-        d_sets_theory.processInference(p, "rels", lemmas);
-        if (d_sets_theory.isInConflict())
+        if (p.getKind() == IMPLIES)
+        {
+          processInference(p[1], p[0], "rels");
+        }
+        else
+        {
+          processInference(p, d_trueNode, "rels");
+        }
+        if (d_state.isInConflict())
         {
           break;
         }
       }
       // if we are still not in conflict, send lemmas
-      if (!d_sets_theory.isInConflict())
+      if (!d_state.isInConflict())
       {
-        d_sets_theory.flushLemmas(lemmas);
+        d_im.flushPendingLemmas();
       }
     }
     d_pending.clear();
   }
 
+  void TheorySetsRels::processInference(Node conc, Node exp, const char* c)
+  {
+    Trace("sets-pinfer") << "Process inference: " << exp << " => " << conc
+                         << std::endl;
+    if (!d_state.isEntailed(exp, true))
+    {
+      Trace("sets-pinfer") << "  must assert as lemma" << std::endl;
+      // we wrap the spurious explanation into a splitting lemma
+      Node lem = NodeManager::currentNM()->mkNode(OR, exp.negate(), conc);
+      d_im.assertInference(lem, d_trueNode, c, 1);
+      return;
+    }
+    // try to assert it as a fact
+    d_im.assertInference(conc, exp, c);
+  }
+
   bool TheorySetsRels::isRelationKind( Kind k ) {
     return k == kind::TRANSPOSE || k == kind::PRODUCT || k == kind::JOIN || k == kind::TCLOSURE;
```

Comment:
Why are Kinds `JOIN_IMAGE, IDEN, APPLY_CONSTRUCTOR` not included in this function?